lib/netext/grpcext/stream.go

```diff
@@ -0,0 +1,135 @@
+package grpcext
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/dynamicpb"
+)
+
+// Stream is the wrapper around the grpc.ClientStream
+// with some handy methods.
+type Stream struct {
+	method           string
+	methodDescriptor protoreflect.MethodDescriptor
+	raw              grpc.ClientStream
+	marshaler        protojson.MarshalOptions
+
+	// this lock is used to prevent concurrent calls to SendMsg & CloseSend
+	// which is not allowed by the grpc.ClientStream
+	// TODO(olegbespalov): consider the better way with the single goroutine
+	// and channels
+	mu *sync.Mutex
+}
+
+// ErrCanceled canceled by client (k6)
+var ErrCanceled = errors.New("canceled by client (k6)")
+
+// ReceiveConverted receives a converted message from the stream
+// if the stream has been closed successfully, it returns io.EOF
+// if the stream has been cancelled, it returns ErrCanceled
+func (s *Stream) ReceiveConverted() (map[string]interface{}, error) {
+	raw, err := s.receive()
+	if err != nil && !errors.Is(err, io.EOF) {
+		return nil, err
+	}
+
+	msg, errConv := s.convert(raw)
+	if errConv != nil {
+		return nil, errConv
+	}
+
+	return msg, err
+}
+
+func (s *Stream) receive() (*dynamicpb.Message, error) {
+	msg := dynamicpb.NewMessage(s.methodDescriptor.Output())
+	err := s.raw.RecvMsg(msg)
+
+	// io.EOF means that the stream has been closed successfully
+	if err == nil || errors.Is(err, io.EOF) {
+		return msg, err
+	}
+
+	sterr := status.Convert(err)
+	if sterr.Code() == codes.Canceled {
+		return nil, ErrCanceled
+	}
+
+	return nil, err
+}
+
+// convert converts the message to the map[string]interface{} format
+// which could be returned to the JS
+// there is a lot of marshaling/unmarshaling here, but if we just pass the dynamic message
+// the default Marshaller would be used, which would strip any zero/default values from the JSON.
+// eg. given this message:
+//
+//	message Point {
+//	   double x = 1;
+//		  double y = 2;
+//		  double z = 3;
+//	}
+//
+// and a value like this:
+// msg := Point{X: 6, Y: 4, Z: 0}
+// would result in JSON output:
+// {"x":6,"y":4}
+// rather than the desired:
+// {"x":6,"y":4,"z":0}
+func (s *Stream) convert(msg *dynamicpb.Message) (map[string]interface{}, error) {
+	// TODO(olegbespalov): add the test that checks that message is not nil
+
+	raw, err := s.marshaler.Marshal(msg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal the message: %w", err)
+	}
+
+	back := make(map[string]interface{})
+
+	err = json.Unmarshal(raw, &back)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal the message: %w", err)
+	}
+
+	return back, err
+}
+
+// CloseSend closes the stream
+func (s *Stream) CloseSend() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
```

Comment:
Seems to me this only gets called through the event loop, so there doesn't really seem to be a reason for this to be locked and unlocked.

The opposite - if the idea is that you should only be calling this from the event loop, a lock here might hide the fact that this is no longer the case and make debugging much harder.

Reply:
Actually, it doesn't call throw the loop since I call that method in the stream's (js module)  `end` method, it's not a big deal to fix it there, but writes are also happening not to throw the loop, but throw the `writeData` routine :thinking:

Reply:
So, I've adjusted it to be called only throw even loop, 665898f4424b1d0b1371e2a56257ff83c2365212 it fixes datarace

Reply:
I think I got confused a bit and then confused you. 

`end` is called only from js code so it is already on the event loop (as js code can only be executed on the event loop). 

`Send` is only called from a `select` loop, so it also technically doesn't need to go on the eventloop. 

I guess the problem is that here the underlying API called by the two should not be called  concurrently.

But that really shouldn't happen on the  event loop as the event loop should not have any networking done on it.

Instead just like in the websockets, `end` should switch the stream in a `closing` state and then of the event loop the actual message should be send. And that send should likely be in that same write loop. This is also what the websocket does. 

Just in the websocket case both are `WriteMessage ` just sends a Close message. 

https://github.com/grafana/xk6-websockets/blob/main/websockets/websockets.go#L628-L639

So in order to not block the event loop on networking I propose that you make the whole closeSend in the writeloop. Don't know if you can reuse the same channel or should just add an additional one.

Reply:
>So in order to not block the event loop on networking I propose that you make the whole closeSend in the writeloop.

done in the latest force push

Reply:
Thanks great job!